### PR TITLE
Fixes #278,Implements "Hi Susi" as hotword command

### LINKED ISSUE #278: Implement "Hi Susi" as hotword wake up commands
**Actual Behaviour**

Currently to talk to Susi, the user needs to tap the mic button. Samsung S6 and Google have apps, that offer voice input by using hotwords/wake up commands. 

**Expected Behaviour**

Please implement "Hi Susi" as a hotword wake up command.

### PR DESCRIPTION
**Pull request by jyoth1raditya**
Fixes issue #278,
Changes: adding CMU's Pocketsphinx to give susi ability to to detect hotwords.
This is a big commit, with changes in 34 files.
This branch is working with voice abilities. No errors are being generated.

### CODE REVIEW

**Comment by chiragw15** [comment]
@jyoth1raditya can you provide an apk for testing?

File: gradle.properties [hunk 1]
```diff
@@ -6,11 +6,16 @@
...
-
- org.gradle.configureondemand=true
+# org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+#
+# When configured, Gradle will run in incubating parallel mode.
+# This option should only be used with decoupled projects. More details, visit
+# http://www.gradle.org/docs/current/userguide/multi_project_builds.html#sec:decoupled_projects
+# org.gradle.parallel=true
+#Sun May 07 15:16:16 IST 2017
+systemProp.http.proxyHost=10.7.0.1
```
Comment (betterclever) [inline gradle.properties]: Do not change project gradle properties to your specific proxy, it will cause build fail.

**Comment by jyoth1raditya** [comment]
@chiragw15, https://www.dropbox.com/s/7xexjzdsvgoocy0/app-debug.apk?dl=0

**Comment by jyoth1raditya** [comment]
Please Help, with this licensing issue
image

**Comment by amitiwary999** [comment]
Have you installed android SDK build-tools 25.0.3?

**Comment by jyoth1raditya** [comment]
@amitiwary99, Can you be more specific

**Comment by betterclever** [comment]
@jyoth1raditya I tried your sample app, not sure if problem is with my
accent but it recognized only 2 times out of 20 times I said "Hi Susi",
Also, after recognition once and a successful stt input, it doesn't
recognize second time.

@amitiwary99 @chiragw15 @mayank408 can you confirm that it is working fine
with you or are you facing similar problems?

**Comment by jyoth1raditya** [comment]
@betterclever, please try the app with speed Internet connection. Coming to accent, it's working fine for me. Please be constructive

**Comment by betterclever** [comment]
@jyoth1raditya I am being constructive only. I just reported some bugs which I found. 😃 
Good work though. This solves a long pending issue.

Also, @Orbiter once recommended that we should have "Susi" as a hotword rather than "Hi Susi". I don't remember exact context of conversation. 
@Orbiter Can you clear that out here "Susi" or "Hi Susi" ?
Thanks.

**Comment by betterclever** [comment]
@jyoth1raditya Also, please Update circle.yml to add latest android sdk and build tools as you updated elsewhere in the project. Else, build shall fail. Thanks.

**Comment by jyoth1raditya** [comment]
@betterclever, Thanks for your suggestions :).
Gonna work on change you suggested

**Comment by chiragw15** [comment]
@jyoth1raditya There is no requirement of internet for hotword detection, right?

**Comment by betterclever** [comment]
@jyoth1raditya Please resolve conflicts. Thanks.

**Comment by jyoth1raditya** [comment]
@chiragw15 Yes, It works offline

**Comment by jyoth1raditya** [comment]
@chiragw15, @daminisatya please review and approve this

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 2]
```diff
@@ -1520,10 +1549,127 @@ private void showToast(String message) {
     public void scrollToEnd(View view) {
         rvChatFeed.smoothScrollToPosition(rvChatFeed.getAdapter().getItemCount() - 1);
     }
+    //code added
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: Add code specific comments.

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 3]
```diff
@@ -1520,10 +1549,127 @@ private void showToast(String message) {
...
+        recognizer.addListener(this);
+
+        /** In your application you might not need to add all those searches.
+         * They are added here for demonstration. You can leave just one.
+         */
+
+        // Create keyword-activation search.
+        recognizer.addKeyphraseSearch(KWS_SEARCH, KEYPHRASE);
+
+
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: Avoid extra empty lines

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 4]
```diff
@@ -59,6 +62,7 @@
 import android.widget.TextView;
 import android.widget.Toast;
 
+
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: Don't add extra empty lines

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 5]
```diff
@@ -154,10 +164,17 @@
...
+    private SpeechRecognizer recognizer;
+    /* Keyword we are looking for to activate menu */
+    private static final String KEYPHRASE = "hi susi";
+    /* Named searches allow to quickly reconfigure the decoder */
+    private static final String KWS_SEARCH = "hi susi";
     private BroadcastReceiver networkStateReceiver;
     private ClientBuilder clientBuilder;
     private Deque<Pair<String, Long>> nonDeliveredMessages = new LinkedList<>();
 
+
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: here also

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 6]
```diff
@@ -1520,10 +1549,127 @@ private void showToast(String message) {
...
+    }
+
+    /**
+     * This callback is called when we stop the recognizer.
+     */
+    @Override
+    public void onResult(Hypothesis hypothesis) {
+
+        if (hypothesis != null) {
+            //makeText(getApplicationContext(), "partial", Toast.LENGTH_SHORT).show();
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: If this code snippet in comments is not useful, remove it

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 7]
```diff
@@ -1520,10 +1549,127 @@ private void showToast(String message) {
...
+        recognizer = SpeechRecognizerSetup.defaultSetup()
+                .setAcousticModel(new File(assetsDir, "en-us-ptm"))
+                .setDictionary(new File(assetsDir, "cmudict-en-us.dict"))
+
+                .setRawLogDir(assetsDir) // To disable logging of raw audio comment out this call (takes a lot of space on the device)
+
+                .getRecognizer();
+        recognizer.addListener(this);
+
+        /** In your application you might not need to add all those searches.
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: May remove this comment

File: build.gradle [hunk 8]
```diff
@@ -8,6 +8,7 @@ buildscript {
         mavenCentral()
     }
     dependencies {
+
```
Comment (chiragw15) [inline build.gradle]: extra empty line

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 9]
```diff
@@ -1520,10 +1549,127 @@ private void showToast(String message) {
...
+    public void onTimeout() {
+        switchSearch(KWS_SEARCH);
+    }
 
     private class computeThread extends Thread {
         public void run() {
             computeOtherMessage();
         }
     }
+
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: extra empty line

**Comment by chiragw15** [comment]
@jyoth1raditya please do the changes. Also, as @betterclever said, @Orbiter can guide you with the exact hotword, whether its "Hi Susi" or just "Susi". Also, if we go by this, the app size almost doubles and becomes 18 mb. Not sure if its recommended. @daminisatya @the-dagger  @mariobehling @Orbiter can test the apk and give feedback on it.  https://www.dropbox.com/s/7xexjzdsvgoocy0/app-debug.apk?dl=0

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 10]
```diff
@@ -154,10 +164,17 @@
     private String reminder;
     private int count = 0;
     private static final String[] id = new String[1];
+
+    private SpeechRecognizer recognizer;
+    /* Keyword we are looking for to activate menu */
+    private static final String KEYPHRASE = "hi susi";
+    /* Named searches allow to quickly reconfigure the decoder */
+    private static final String KWS_SEARCH = "hi susi";
```
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: @chiragw15 @jyoth1raditya @betterclever After testing the app, I found that it only recognized 2 to 3 times out of ten. The reason I found is the change in accents. On testing with other speech to text engines when we say "Hi Susi" It actually interprets it as "Hi Sushi". Due to which the feature does not work at most times. The probable solution could be to have a list of Strings of probable phrases similar to "Hi Susi" and then match it.

**Comment by mariobehling** [comment]
Please resolve merge conflicts.

**Comment by mariobehling** [comment]
Avoid "Fixes #278" in the commit message itself. It is sufficient to have this info in the description.

**Comment by jyoth1raditya** [comment]
@mariobehling , Resolved the merge conflicts

**Comment by chiragw15** [comment]
@mariobehling do not merge this. We are also testing snowboy for hotword detection. We will compare result from both Snowboy and CMU PocketSphinx and will pick the better one. If this is merged now, it will be very difficult to revert it.

**Comment by amitiwary999** [comment]
@chiragw15 we are using snowboy so this pr is needed or not?

**Comment by mariobehling** [comment]
I am closing this PR due to inactivity. Please make a new PR to follow up if you decide to continue on this.

**Pull request closed without merging**